dll/win32/framedyn/wbemtime.cpp

```diff
@@ -0,0 +1,683 @@
+{
+    if (IsOk() && uAdd.IsOk())
+    {
+        m_Time += uAdd.m_Time;
+    }
+    else
+    {
+        m_Time = INVALID_TIME;
+    }
+
+    return *this;
+}
+
+// subtract operator (�) subtracts a time span from the object on which the method is executed. The result is a new time span that contains the difference in time.
+WBEMTimeSpan WBEMTimeSpan::operator-(const WBEMTimeSpan &uSub) const
+{
+    WBEMTimeSpan rt;
+
+    if (IsOk() && uSub.IsOk() && m_Time >= uSub.m_Time)
+    {
+        rt.m_Time = m_Time - uSub.m_Time;
+    }
+
+    return rt;
+}
+
+// subtract and assign operator (�=).
+const WBEMTimeSpan& WBEMTimeSpan::operator-=(const WBEMTimeSpan &uSub)
+{
+    if (IsOk() && uSub.IsOk() && m_Time >= uSub.m_Time)
+    {
+        m_Time -= uSub.m_Time;
+    }
+    else
+    {
+        m_Time = INVALID_TIME;
+    }
+
+    return *this;
+}
+
+// Converts a BSTR time interval value to a WBEMTimeSpan object in CIM date and time format.
+const WBEMTimeSpan& WBEMTimeSpan::operator=(const BSTR bstrDMTFFormat)
+{
+    m_Time = INVALID_TIME;
+
+    if (bstrDMTFFormat != NULL && wcslen(bstrDMTFFormat) == 25 && bstrDMTFFormat[14] == L'.' && bstrDMTFFormat[21] == L':')
+    {
+        int iDays, iHours, iMinutes, iSeconds, iUSec;
+        if (swscanf(bstrDMTFFormat, L"%8d%2d%2d%2d.%6d:000", &iDays, &iHours, &iMinutes, &iSeconds, &iUSec) == 5)
+        {
+            WBEMTimeSpan t(iDays, iHours, iMinutes, iSeconds, 0, iUSec, 0);
+            m_Time = t.m_Time;
+        }
+    }
+
+    return *this;
+}
+
+BOOL WBEMTimeSpan::operator==(const WBEMTimeSpan &a) const
+{
+    return m_Time == a.m_Time;
+}
+
+BOOL WBEMTimeSpan::operator!=(const WBEMTimeSpan &a) const
+{
+    return m_Time != a.m_Time;
+}
+
+BOOL WBEMTimeSpan::operator<(const WBEMTimeSpan &a) const
+{
+    return m_Time < a.m_Time;
+}
+
+BOOL WBEMTimeSpan::operator<=(const WBEMTimeSpan &a) const
+{
+    return m_Time <= a.m_Time;
+}
+
+BOOL WBEMTimeSpan::operator>(const WBEMTimeSpan &a) const
+{
+    return m_Time > a.m_Time;
+}
+
+BOOL WBEMTimeSpan::operator>=(const WBEMTimeSpan &a) const
+{
+    return m_Time >= a.m_Time;
+}
+
+// gets the time span as a BSTR in Date and Time format.
+BSTR WBEMTimeSpan::GetBSTR(void) const throw (CHeap_Exception)
+{
+    if (!IsOk())
+    {
+        return NULL;
+    }
+
+    ULONGLONG totalSeconds = m_Time / SECONDS_OFFSET;
+    ULONGLONG iDays = totalSeconds / SECONDS_PER_DAY;
+    int iHours = (totalSeconds % SECONDS_PER_DAY) / SECONDS_PER_HOUR;
+    int iMinutes = (totalSeconds % SECONDS_PER_HOUR) / SECONDS_PER_MINUTE;
+    int iSeconds = totalSeconds % SECONDS_PER_MINUTE;
+    int iUSec = (m_Time % SECONDS_OFFSET) / MICROSECONDS_OFFSET;
+
+    BSTR ret = (BSTR)SysAllocStringLen(NULL, 26);
+    if (StringCchPrintfW(ret, 26, L"%08I64i%02d%02d%02d.%06ld:000", iDays, iHours, iMinutes, iSeconds, iUSec) != S_OK)
+    {
+        return NULL;
+    }
+    return ret;
+}
+
+bool WBEMTimeSpan::IsOk() const
+{
+    return m_Time != INVALID_TIME ? TRUE : FALSE;
```

Comment:
`bool` takes C++ specific values (true/false) and not TRUE/FALSE.

Reply:
How is this

Reply:
No. I meant the `WBEMTimeSpan::IsOk` function is declared as a C++ boolean specific data type and returns whether false or true depending on the coder's purpose for its logic implementation. `TRUE` and `FALSE` are `BOOL` Win32 return values which they're reserved.

To sum this all up, just change the one-liner with this:

```C
return m_Time != INVALID_TIME ? true : false;
```

Reply:
`? true : false` is an anti-pattern. Just use `return m_Time != INVALID_TIME;`